plugins/org.eclipse.emf.ecoretools.ale.core/src/org/eclipse/emf/ecoretools/ale/core/env/impl/MadeUpAleEnvironment.java

```diff
@@ -0,0 +1,77 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Inria and Obeo.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Inria - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.emf.ecoretools.ale.core.env.impl;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+
+import org.eclipse.acceleo.query.runtime.IQueryEnvironment;
+import org.eclipse.emf.ecore.EPackage;
+import org.eclipse.emf.ecoretools.ale.core.env.IAleEnvironment;
+import org.eclipse.emf.ecoretools.ale.core.env.IBehaviors;
+import org.eclipse.emf.ecoretools.ale.core.parser.ParsedFile;
+import org.eclipse.emf.ecoretools.ale.implementation.ModelUnit;
+
+/**
+ * An ALE environment which is made up of explicit metamodels and behaviors.
+ * <p>
+ * This class is not supposed to be instantiated by clients, call 
+ * {@link IAleEnvironment#fromRaw(IQueryEnvironment, Collection, Collection)}
+ * instead.
+ */
+public class MadeUpAleEnvironment extends AbstractAleEnvironment {
```

Comment:
The terms "MadeUp" in the name of this class doesn't ring a bell to me 
we should try to find a better name, 

the distinction with the _InMemoryAleEnvironment_ needs some clarification ...

may be something related to the acceleo query environment ?

Reply:
True. Basically:

- `InMemory` is an ALE environment that only exists at runtime and that is dynamically defined (and loaded) from a set of paths to _.ecore_ and _.ale_ files,
- `MadeUp` is not loaded from files but rather made up of a set of `EPackages` and `BehavioredClasses` that are explicitely specified by the caller.

I don't have good alternatives for `MadeUp`. I thought of `ComprehensiveAleEnvironment` or `ExplicitAleEnvironment` but I don't like them either.

Now that I rethink about it, `ImmutableAleEnvironment` may fit. Instances won't actually be immutables because none of its arguments is, but the class is still aimed at representing immutable environments built from a comprehensive set of EMF packages and ALE behaviors.

Reply:
- `InMemory` has been renamed as `PathsBasedAleEnvironment`
 - `MadeUp` has been renamed as `RawEnvironmentBasedAleEnvironment`

Class names are now aligned with the name of the factory methods clients are supposed to use to instantiate them: `IAleEnvironment.fromPaths` and `IAleEnvironment.fromRaw`.

Clients are not supposed to be aware of these class names, but I think they now are understandable if clients find them (e.g. in the debugger).